http://git.kernel.org/?p=linux/kernel/git/nico/orion.git;a=commitdiff;h=31481adbabd1a6d66af40f1508531af3bed6f2c3

From: Zintis Pētersons <zintis.petersons@abcsolutions.lv>
Subject: [PATCH] Kirkwood: initialize PCIE1 for QNAP TS-419P+
To: nico@fluxnic.net
Cc: linux-arm-kernel@lists.infradead.org

Initialize PCIE1 on the 6282-based QNAP TS-419P+ since it has a
Marvell 9125 SATA chip on each PCI bus.

Signed-off-by: Zintis Pētersons <zintis.petersons@abcsolutions.lv>

diff --git a/arch/arm/mach-kirkwood/ts41x-setup.c b/arch/arm/mach-kirkwood/ts41x-setup.c
index 9a44029..11f23c3 100644
--- a/arch/arm/mach-kirkwood/ts41x-setup.c
+++ b/arch/arm/mach-kirkwood/ts41x-setup.c
@@ -251,6 +251,8 @@ static void __init qnap_ts41x_init(void)
 
 static int __init ts41x_pci_init(void)
 {
+	u32 dev, rev;
+
 	if (machine_is_ts41x()) {
 		/*
 		 * Without this explicit reset, the PCIe SATA controller
@@ -259,7 +261,11 @@ static int __init ts41x_pci_init(void)
 		 */
 		orion_pcie_reset((void __iomem *)PCIE_VIRT_BASE);
 
-		kirkwood_pcie_init(KW_PCIE0);
+		kirkwood_pcie_id(&dev, &rev);
+		if (dev == MV88F6282_DEV_ID)
+			kirkwood_pcie_init(KW_PCIE1 | KW_PCIE0);
+		else
+			kirkwood_pcie_init(KW_PCIE0);
 	}
 
    return 0;
